lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp

```diff
@@ -0,0 +1,218 @@
+
+    // add injection of state loaded in SPGW state manager
+    std::string homedir = std::string(getpwuid(getuid())->pw_dir);
+    name_of_ue_samples  = load_file_into_vector_of_line_content(
+        homedir + "/" + DEFAULT_SPGW_CONTEXT_DATA_PATH,
+        homedir + "/" +
+            "magma/lte/gateway/c/core/oai/test/spgw_task/data/"
+            "data_list.txt");
+    mock_read_spgw_ue_state_db(name_of_ue_samples);
+
+    std::this_thread::sleep_for(
+        std::chrono::milliseconds(END_OF_TEST_SLEEP_MS));
+    std::cout << "Setup done" << std::endl;
+  }
+
+  virtual void TearDown() {
+    send_terminate_message_fatal(&task_zmq_ctx_main_spgw);
+    destroy_task_context(&task_zmq_ctx_main_spgw);
+    itti_free_desc_threads();
+
+    free_mme_config(&mme_config);
+    free_spgw_config(&spgw_config);
+
+    // Sleep to ensure that messages are received and contexts are released
+    std::this_thread::sleep_for(
+        std::chrono::milliseconds(END_OF_TEST_SLEEP_MS));
+  }
+
+ protected:
+  std::shared_ptr<MockMmeAppHandler> mme_app_handler;
+  std::string test_imsi_str          = "001010000000002";
+  unsigned long long int test_imsi64 = 1010000000002;
+  plmn_t test_plmn                   = {.mcc_digit2 = 0,
+                      .mcc_digit1 = 0,
+                      .mnc_digit3 = 0x00,
+                      .mcc_digit3 = 0,
+                      .mnc_digit2 = 0,
+                      .mnc_digit1 = 0};
+  bearer_context_to_be_created_t sample_default_bearer_context = {
+      .eps_bearer_id    = 5,
+      .bearer_level_qos = {
+          .pci = 1,
+          .pl  = 15,
+          .qci = 9,
+          .gbr = {},
+      }};
+  int test_mme_s11_teid = 14;
+  int test_bearer_index = 5;
+  in_addr_t test_ue_ip  = 0x05030201;
+  in_addr_t test_ue_ip2 = 0x0d80a8c0;
+  std::vector<std::string> name_of_ue_samples;
+};
+
+TEST_F(SPGWAppInjectedStateProcedureTest, TestCreateSessionPCEFFailure) {
+  spgw_state_t* spgw_state = get_spgw_state(false);
+
+  spgw_ue_context_t* ue_context_p = spgw_get_ue_context(test_imsi64);
+
+  teid_t ue_sgw_teid =
+      LIST_FIRST(&ue_context_p->sgw_s11_teid_list)->sgw_s11_teid;
+
+  s_plus_p_gw_eps_bearer_context_information_t* spgw_eps_bearer_ctxt_info_p =
+      sgw_cm_get_spgw_context(ue_sgw_teid);
+
+  sgw_eps_bearer_ctxt_t* eps_bearer_ctxt_p = sgw_cm_get_eps_bearer_entry(
+      &spgw_eps_bearer_ctxt_info_p->sgw_eps_bearer_context_information
+           .pdn_connection,
+      DEFAULT_EPS_BEARER_ID);
+
+  // send an IP alloc response to SPGW
+  itti_ip_allocation_response_t test_ip_alloc_resp = {};
+  fill_ip_allocation_response(
+      &test_ip_alloc_resp, SGI_STATUS_OK, ue_sgw_teid, DEFAULT_EPS_BEARER_ID,
+      test_ue_ip2, DEFAULT_VLAN);
+  status_code_e ip_alloc_rc = sgw_handle_ip_allocation_rsp(
+      spgw_state, &test_ip_alloc_resp, test_imsi64);
+
+  // check if IP address is allocated after this message is done
+  ASSERT_TRUE(eps_bearer_ctxt_p->paa.ipv4_address.s_addr == test_ue_ip2);
```

Comment:
Can you check that this assert is false before the call to `sgw_handle_ip_allocation_rsp`?

Reply:
@ssanadhya Correct me if I am wrong. Since the state for the test_imsi64 captured after Attach was completed, SPGW task should not expect an IP allocation response for this IMSI, then we don't need to check that this assert is false anymore. Is this correct?

Reply:
@thhung , that's right. In fact, since the state was captured after Attach was complete, the TestPCEFFailure is not applicable here, since session is already created. You can add tests for 
- Delete Session
- Release Bearer
- Secondary PDN connections 
- Dedicated Bearer activation/deactivation